letsencrypt/client/reverter.py

```diff
@@ -0,0 +1,433 @@
+            op_fd = open(filepath, 'r+')
+            lines = op_fd.read().splitlines()
+        else:
+            lines = []
+            op_fd = open(filepath, 'w')
+
+        return op_fd, lines
+
+    def _recover_checkpoint(self, cp_dir):
+        """Recover a specific checkpoint.
+
+        Recover a specific checkpoint provided by cp_dir
+        Note: this function does not reload augeas.
+
+        :param str cp_dir: checkpoint directory file path
+
+        :raises errors.LetsEncryptReverterError: If unable to recover checkpoint
+
+        """
+        if os.path.isfile(os.path.join(cp_dir, "FILEPATHS")):
+            try:
+                with open(os.path.join(cp_dir, "FILEPATHS")) as paths_fd:
+                    filepaths = paths_fd.read().splitlines()
+                    for idx, path in enumerate(filepaths):
+                        shutil.copy2(os.path.join(
+                            cp_dir,
+                            os.path.basename(path) + '_' + str(idx)), path)
+            except (IOError, OSError):
+                # This file is required in all checkpoints.
+                logging.error("Unable to recover files from %s", cp_dir)
+                raise errors.LetsEncryptReverterError(
+                    "Unable to recover files from %s" % cp_dir)
+
+        # Remove any newly added files if they exist
+        self._remove_contained_files(os.path.join(cp_dir, "NEW_FILES"))
+
+        try:
+            shutil.rmtree(cp_dir)
+        except OSError:
+            logging.error("Unable to remove directory: %s", cp_dir)
+            raise errors.LetsEncryptReverterError(
+                "Unable to remove directory: %s" % cp_dir)
+
+    def _check_tempfile_saves(self, save_files):  # pylint: disable=no-self-use
+        """Verify save isn't overwriting any temporary files.
+
+        :param set save_files: Set of files about to be saved.
+
+        :raises :class:`letsencrypt.client.errors.LetsEncryptReverterError`:
+            when save is attempting to overwrite a temporary file.
+
+        """
+        protected_files = []
+
+        # Get temp modified files
+        temp_path = os.path.join(self.direc['temp'], "FILEPATHS")
+        if os.path.isfile(temp_path):
+            with open(temp_path, 'r') as protected_fd:
+                protected_files.extend(protected_fd.read().splitlines())
+
+        # Get temp new files
+        new_path = os.path.join(self.direc['temp'], "NEW_FILES")
+        if os.path.isfile(new_path):
+            with open(new_path, 'r') as protected_fd:
+                protected_files.extend(protected_fd.read().splitlines())
+
+        # Verify no save_file is in protected_files
+        for filename in protected_files:
+            if filename in save_files:
+                raise errors.LetsEncryptReverterError(
+                    "Attempting to overwrite challenge "
+                    "file - %s" % filename)
+
+    # pylint: disable=no-self-use, anomalous-backslash-in-string
```

Comment:
https://github.com/letsencrypt/lets-encrypt-preview/pull/185/files#r23503203

`anomalous-backslash-in-string` is not necessary if `r"""` is used.